### PR TITLE
[FLOC-3891] Helpers for making files and directories 

### PR DESCRIPTION
**Pull request by jml**
Adds new helpers in `flocker.common`:

* `make_file` – create a file with content and permissions
* `make_directory` – create a directory, or do nothing when a directory already exists

Similar code is repeated a lot throughout the code base, so I thought it might be nice to have something more intent-revealing.

Further:

* generators for Unix permissions
* matcher for Unix permissions
* `make_temporary_file` now takes optional permissions

This answers an in-person question asked by @avg-I.

### CODE REVIEW

File: flocker/common/test/test_filepath.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        """
+        If the directory exists, ``make_directory`` does nothing.
+        """
+        path = self.make_temporary_directory()
+        make_directory(path)
+        self.assertThat(path, dir_exists())
+
+    def test_make_directory_file_exists(self):
+        """
+        If the file exists, ``make_directory`` does something.
```
Comment (sarum90) [inline flocker/common/test/test_filepath.py]: :grey_question: s/does something/raises an exception/

**Comment by sarum90** [comment]
Looks good to me, thanks!

File: flocker/common/_filepath.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    """
+    Create a file with given content and permissions.
+
+    :param FilePath path: Path to create the file.
+    :param str content: Content to write to the file. If not specified,
+    :param int permissions: Unix file permissions to be passed to ``chmod``.
+    """
+    path.setContent(content)
+    if permissions is not None:
+        path.chmod(permissions)
```
Comment (avg-I) [inline flocker/common/_filepath.py]: I wonder if it should be `create` -> `chmod` -> `setContent` in case the content is sensitive.

Comment (jml) [inline flocker/common/_filepath.py]: I tried this, and became sad: https://github.com/twisted/twisted/blob/trunk/twisted/python/filepath.py#L1528-L1536

File: flocker/common/test/test_filepath.py (hunk 1 above)
Comment (jml) [inline flocker/common/test/test_filepath.py]: Fixed.

File: flocker/common/_filepath.py (hunk 2 above)
Comment (avg-I) [inline flocker/common/_filepath.py]: Argh, that code is sad indeed.

**Pull request merged**